src/VisualStudio/Core/Def/ProjectSystem/CPS/IWorkspaceProjectContextFactory.cs

```diff
@@ -42,5 +38,61 @@ internal interface IWorkspaceProjectContextFactory
             string? binOutputPath,
             string? assemblyName,
             CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Creates and initializes a new project and returns a <see
+        /// cref="IWorkspaceProjectContext"/> to lazily initialize the properties and items for the
+        /// project.  This method guarantees that either the project is added (and the returned task
+        /// completes) or cancellation is observed and no project is added.
+        /// </summary>
+        /// <param name="id">Project guid</param>
+        /// <param name="uniqueName">
+        /// Unique across the entire solution for the life of the solution. 
+        /// This will be unique across regardless of whether projects are added or renamed 
+        /// to match this project's original name.
+        /// </param>
+        /// <param name="data">Providers access to msbuild evaluation data for the project.</param>
+        /// <param name="hostObject">The IVsHierarchy for the project; this is used to track linked files across multiple projects when determining contexts.</param>
+        /// <exception cref="InvalidOperationException">A required property or item is not present in <see cref="EvaluationData"/> or has invalid value.</exception>
+        Task<IWorkspaceProjectContext> CreateProjectContextAsync(Guid id, string uniqueName, string languageName, EvaluationData data, object? hostObject, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Names of msbuild properties whose values that <see cref="CreateProjectContextAsync(Guid, string, string, EvaluationData, object?, CancellationToken)"/> will receive via <see cref="EvaluationData"/>.
+        /// </summary>
+        ImmutableArray<string> EvaluationPropertyNames { get; }
+    }
+
+    internal abstract class EvaluationData
+    {
+        /// <summary>
+        /// Returns the value of property of the specified <paramref name="name"/>.
+        /// </summary>
+        /// <returns>
+        /// Returns empty string if the property is not set.
+        /// </returns>
+        /// <exception cref="ArgumentException">
+        /// The <paramref name="name"/> is not listed in <see cref="IWorkspaceProjectContextFactory.EvaluationPropertyNames"/>
+        /// </exception>
+        public abstract string GetPropertyValue(string name);
+
+        public string GetRequiredPropertyValue(string name)
+        {
+            var value = GetPropertyValue(name);
+
+            if (string.IsNullOrEmpty(value))
+                throw new InvalidOperationException($"Property '{name}' is required.");
+
+            return value!;
```

Comment:
I think we have RoslynString.IsNullOrEmpty that's properly annotated here, but I'm a bit surprised this needed the ! since the GetPropertyValue should have been non-null to start with?